src/lib/xmlparse.rs

```diff
@@ -9332,41 +9333,43 @@ unsafe extern "C" fn poolDestroy(mut pool: *mut STRING_POOL) {
     }
 }
 
-unsafe extern "C" fn poolAppend(
-    mut pool: *mut STRING_POOL,
-    mut enc: &ENCODING,
-    mut ptr: *const c_char,
-    mut end: *const c_char,
-) -> *mut XML_Char {
-    if (*pool).ptr.is_null() && poolGrow(pool) == 0 {
-        return NULL as *mut XML_Char;
-    }
-    loop {
-        let convert_res: super::xmltok::XML_Convert_Result = XmlConvert!(
-            enc,
-            &mut ptr,
-            end,
-            &mut (*pool).ptr as *mut *mut _ as *mut *mut ICHAR,
-            (*pool).end as *mut ICHAR,
-        );
-        if convert_res == super::xmltok::XML_CONVERT_COMPLETED
-            || convert_res == super::xmltok::XML_CONVERT_INPUT_INCOMPLETE
-        {
-            break;
-        }
-        if poolGrow(pool) == 0 {
+impl STRING_POOL {
+    unsafe fn append(
+        &mut self,
+        enc: &ENCODING,
+        mut ptr: *const c_char,
+        end: *const c_char,
+    ) -> *mut XML_Char {
+        if self.ptr.is_null() && self.grow() == 0 {
             return NULL as *mut XML_Char;
         }
+        loop {
+            let convert_res: super::xmltok::XML_Convert_Result = XmlConvert!(
+                enc,
+                &mut ptr,
+                end,
+                &mut self.ptr as *mut *mut _ as *mut *mut ICHAR,
+                self.end as *mut ICHAR,
+            );
+            if convert_res == super::xmltok::XML_CONVERT_COMPLETED
+                || convert_res == super::xmltok::XML_CONVERT_INPUT_INCOMPLETE
+            {
+                break;
+            }
+            if self.grow() == 0 {
+                return NULL as *mut XML_Char;
+            }
+        }
+        self.start
     }
-    return (*pool).start;
 }
 
 unsafe extern "C" fn poolCopyString(
```

Comment:
Shouldn't this be a method too?

Reply:
There's a bunch more to refactor, this is just the first pass.